### PR TITLE
Word size suffix and BigNumber support for bin oct hex literals

### PR DESCRIPTION
**Pull request by clnhlzmn**
I am working on adding a word size suffix for binary, octal, and hex literals like `0xffffi32`.

My plan is to parse literals with prefixes `0b`, `0c`, and `0x` as unsigned by default. If you want to parse them as signed you have to specify a word size such as `i16` or `i45`. 

If you're using `Number` as your default type then the max word size is `i53`. If you're using `BigNumber` as default then I am not sure what the max word size would be but it would be larger than 53.

The format functions such as `bin`, `oct`, and `hex` will assume unsigned by default. If you want to format the number as signed you will specify a word size such as `bin(n, {'wordSize':'32'})`.

### CODE REVIEW

**Comment by clnhlzmn** [comment]
Extending the word size beyond 32 bits (when type is Number) will require changing the implementation of bitwise operators if we expect something like `mathjs.evaluate('0xffffffffffi40 & 0xff00000000')`. Currently that expression is `0` when it should be `0xff00000000`.

**Comment by clnhlzmn** [comment]
I just realized that decimal.js doesn't do bitwise operations. That's a bummer.

**Comment by josdejong** [comment]
I just had a short look at your progress, looks good so far 👍 . 

As for the bitwise operations with BigNumber: I think you can use https://github.com/josdejong/mathjs/blob/develop/src/utils/bignumber/bitwise.js ?

**Comment by clnhlzmn** [comment]
> As for the bitwise operations with BigNumber: I think you can use ...

I didn't  notice that. Very cool.

**Comment by clnhlzmn** [comment]
Since Js integers are limited to 53 bits I have limited the non decimal literals to 53 bits so a user isn't confused by an expression like `0x1000000000000000 === 0xfffffffffffffff` being true. But I wonder if it isn't necessary to impose that limit since plain Js happily allows you to input literals that don't end up being exactly equal to what you think. Not sure if that's clear. Wondering what you think about that @josdejong.

**Comment by clnhlzmn** [comment]
# An update on what I've done here:

Please let me know what you think @josdejong.

## literals

The default is for literals to be parsed as unsigned:

    0xffffffff == 4294967295

If you want to enter a signed literal you need to specify a word size:

    0xffffffffi32 == -1
    0x8000000000000000i64 == -9223372036854776000

The word size suffix is only for signed values. I didn't add an unsigned suffix like `uNN` because the default is unsigned. If the value is larger than can fit into a word of the specified size then an error is thrown:

    0x100i8 //Error: String "0x100" is out of range

The size suffix can be any positive number

    0b100i3 == -4

## formatting

I added `'bin'`, `'oct'`, and `'hex'` values for the `notation` option and a `wordSize` option to the `format` function:

    math.format(128, { notation: 'hex' }) == '0x80'

The default formatting for negative values doesn't assume a word size and instead just prepends a `-` (like plain js):

    math.format(-128, { notation: 'hex' }) == '-0x80'

If you specify `wordSize` then `format` formats negative values as if they're two's complement. The `wordSize` option only has an effect if used with one of the `bin`, `oct`, or `hex` notations:

    math.format(-128, { notation: 'hex', wordSize: 8 }) == '0x80'

`wordSize` can be any positive integer.

You can format non integer values in an alternative base:

    math.format(1.25, { notation: 'hex' }) == '0b1.01'

If the value passed to `format` is too large to fit into a word of the specified size an error is thrown:

    math.format(128, { notation: 'hex', wordSize: 8 }) //Error: Value must be in range [-2^7, 2^7-1]
    
The functions `bin`, `oct`, and `hex` are just shorthand for calling `format` with the `notation` option set to `'bin'`, `'oct'`, or `'hex'`. They also take an optional `wordSize` argument:

    bin(128) == '0b10000000'
    oct(-128) == '-0o200'
    hex(-128, 8) == '0x80'
    
## BigNumber

These changes should now work with `BigNumber`.

## notes

* There is no longer a restriction on the size of literals. If you're using `number` you can write literals that will exceed the available precision (53 bits):

      0xffffffffffffffffi64 != -1

  I did this to be consistent with plain js which also allows you to write literals that exceed the available precision. Perhaps there should be a warning?

  If you use `BigNumber` that isn't an issue:

      bignumber('0xffffffffffffffffi64') == -1

* If you're using `number` bitwise operations are further limited to 32 bit signed operands, but `BigNumber` is not restricted in this way:

      0xffffffff << 1 == -2
      bignumber('0xffffffff << 1') == 0x1fffffffe
    
  I think it would be a good idea to extend bitwise operations with `number` to the full 53 bit precision.

**Comment by josdejong** [comment]
Thanks for the update Clinton! Looks promising. I'll look into it soon.

**Comment by josdejong** [comment]
wow, this is impressive, Colin! It looks really solid and well thought through. The code is really straightforward, easy to understand, and well tested.

Some feedbacks:

1.  I do like the default of having non-signed numbers when no literal like `i32` is provided. I think this is a breaking change, right? So we'll want to publish this as a new major version increment (which is fine).

2.  It makes sense to me to throw an exception when trying to parse a literal which cannot be represented, but this issue is not unique to these literals: you can enter `math.evaluate('2e500')`, which will not return what you expect. It is a generic problem which we need to think through separately I think (we'll have to think through what to do with decimals which cannot represented exactly, etc).

3.  > I think it would be a good idea to extend bitwise operations with number to the full 53 bit precision.
    
    Yes that makes sense. Do you think it's best do address this in this PR, or in a separate PR?

4.  It would be great if we could also parse decimal literals like `0b1.01` with the expression parser, so we can actually consume the output of `format`, `hex`, `bin`, `oct` again.

5.  We can now parse hex/bin/oct literals unambiguously like `0xffffffffi32`. Is there a way to format a number unambiguously too? I.e. have `math.format` or `math.hex` output `0xffffffffi32`? That would make this functionality symmetric, so the formatted output could be fed into the expression parser again unambiguously. What do you think?

6.  Currently, if you make a mistake by forgetting to type the `i`, an expression is evaluated as an implicit multiplication. So for example `0xff32` is evaluated as `0xff * 32` and returns 65330. It is valid syntax but I think this can be an ugly pitfall. Maybe we should simply not allow implicit multiplication here, by throwing an exception when the literal is followed by a digit?

7.  I think we still need some additional documentation to explain those new features :sunglasses:

File: src/function/string/bin.js [hunk 1]
```diff
@@ -20,4 +23,13 @@ import { createBaseFormatterFactory } from './baseUtils.js'
...
  * @return {string}         The formatted value
  */
-export const createBin = createBaseFormatterFactory('bin', 2)
+export const createBin = factory(name, dependencies, ({ typed, format }) => {
+  return typed(name, {
+    'number | BigNumber': function (n) {
+      return format(n, { notation: 'bin' })
+    },
+    'number | BigNumber, number': function (n, wordSize) {
+      return format(n, { notation: 'bin', wordSize: wordSize })
```
Comment (josdejong) [inline src/function/string/bin.js:32]: Very minor remark: you could simplify

```js
      return format(n, { notation: 'bin', wordSize: wordSize })
```
to
```js
      return format(n, { notation: 'bin', wordSize })
```

**Comment by clnhlzmn** [comment]
> wow, this is impressive, Colin! It looks really solid and well thought through. The code is really straightforward, easy to understand, and well tested.

Thanks!

> Some feedbacks:
> 
> 1. I do like the default of having non-signed numbers when no literal like `i32` is provided. I think this is a breaking change, right? So we'll want to publish this as a new major version increment (which is fine).

Yes this is a breaking change.

> 2. It makes sense to me to throw an exception when trying to parse a literal which cannot be represented, but this issue is not unique to these literals: you can enter `math.evaluate('2e500')`, which will not return what you expect. It is a generic problem which we need to think through separately I think (we'll have to think through what to do with decimals which cannot represented exactly, etc).

It makes sense to work on this as a separate problem.

> 3. > I think it would be a good idea to extend bitwise operations with number to the full 53 bit precision.
>
>  Yes that makes sense. Do you think it's best do address this in this PR, or in a separate PR?

I think it makes sense to address it separately.

> 4. It would be great if we could also parse decimal literals like `0b1.01` with the expression parser, so we can actually consume the output of `format`, `hex`, `bin`, `oct` again.

I agree. I don't think that js supports this natively (though decimal.js does) so I'll have to think about how to implement it for number. Perhaps this can be added later.

> 5. We can now parse hex/bin/oct literals unambiguously like `0xffffffffi32`. Is there a way to format a number unambiguously too? I.e. have `math.format` or `math.hex` output `0xffffffffi32`? That would make this functionality symmetric, so the formatted output could be fed into the expression parser again unambiguously. What do you think?

That makes sense. I think it makes sense that if you provide the `wordSize` option to `format` then we can append the suffix to the output. Then everything is symmetrical. I will add this.

> 6. Currently, if you make a mistake by forgetting to type the `i`, an expression is evaluated as an implicit multiplication. So for example `0xff32` is evaluated as `0xff * 32` and returns 65330. It is valid syntax but I think this can be an ugly pitfall. Maybe we should simply not allow implicit multiplication here, by throwing an exception when the literal is followed by a digit?

I think this is working as expected. What's happening is that `0xff32` is just getting parsed as a single number `65330` not as a multiplication. `0xff32 == 65330` and `0xff * 32 == 8160`. If you try something like `0b1132` where some of the digits are illegal for the type of literal it doesn't parse as a multiplication, it throws an error.

> 7. I think we still need some additional documentation to explain those new features 😎

I will update the documentation. I wanted to get your opinion first.

**Comment by josdejong** [comment]
1) :thumbsup:
2) :thumbsup:
3) :thumbsup:
4) :thumbsup: would be nice, it would be the "opposite" of being able to do `math.format(1.25, { notation: 'bin' })` which is now supported, and which suggests mathjs can work with decimals for hex/bin/oct.
5) :thumbsup:
6) Aaaah, of course, my bad you're right! :joy:
7) :thumbsup:

**Comment by josdejong** [comment]
Thanks for the latest updates, I'll look into the PR again soon.

**Comment by clnhlzmn** [comment]
👍

**Comment by josdejong** [comment]
I've played around with your PR again Colin, it looks really solid and consistent! I like it 👍 

Will merge your PR now in a `v9` branch since it contains a breaking change. I expect to publish v9 soon but will await a little bit to see if there are other breaking changes that we would like to publish in v9 too.

**Comment by josdejong** [comment]
Published now in `v9.0.0` 🎉 . Thanks again.

**Pull request merged**